### PR TITLE
Adding a new command equivalent to linux shell 'whoami'

### PR DESCRIPTION
**Pull request by asifhuawei**
This command will help user to get the information related to the Org and Space they have targeted and the role associated with them.

Output : 

```
$ cf user-info
Getting user information......
User    Org       Space       Role
asif   asifOrg   asifSpace   ORG MANAGER, SPACE MANAGER, SPACE DEVELOPER
```

Issue : http://rnd-github.huawei.com/cloudfoundry/cli/issues/16

### CODE REVIEW

**Comment by SrinivasChilveri** [comment]
@asifhuawei 
I think we can do requirements validation instead of 

if cmd.config.SpaceFields().Name == "" { 
 44 +		if cmd.config.OrganizationFields().Name == "" { 
 45 +			message := fmt.Sprintf(T("No org and space targeted, use '{{.Command}}' to target an org and space", map[string]interface{}{"Command": terminal.CommandColor(cf.Name() + " target -o ORG -s SPACE")})) 
 46 +			cmd.ui.Failed(message) 
 47 +			return 
 48 +		} else { 
 49 +			message := fmt.Sprintf(T("No space targeted, use '{{.Command}}' to target a space", map[string]interface{}{"Command": terminal.CommandColor("cf target -s")})) 
 50 +			cmd.ui.Failed(message) 
 51 +			return 
 52 +		} 
 53 +	} 

as below  if we need login & space requirements...

reqs = []requirements.Requirement{
		requirementsFactory.NewLoginRequirement(),
		requirementsFactory.NewTargetedSpaceRequirement(),
		cmd.appReq,
	}

Even i feel we can think of implementing by sending the user name to CC & get the rols instead of getting all the users of all the org & space roles(need to check the feasibility in CC).

Need to discuss in the team on this.

**Comment by SrinivasChilveri** [comment]
I think  in Context("when logged in and targeted with proper orgs and space", func() { 
 58 +		BeforeEach(func() { 

we are doing wrong org fileds setting to space& spacefilds to org .
Instead of so many lines & doing copy pastemistake better we do 
as below
config.SetOrganizationFields(org )
config.SetSpaceFields(space)

configRepo.SetSpaceFields(models.SpaceFields{ 
 87 +				Name: "Org1", 
 88 +				Guid: "org1-guid", 
 89 +			}) 
 90 + 
 91 +			configRepo.SetOrganizationFields(models.OrganizationFields{ 
 92 +				Name: "Space1", 
 93 +				Guid: "space1-guid", 
 94 +			})

File: cf/commands/user/user-info.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+			users, apiErr = cmd.userRepo.ListUsersInOrgForRole(cmd.config.OrganizationFields().Guid, role)
+		}
+
+		for _, user := range users {
+			if cmd.config.Username() == user.Username {
+				retrivedRoles = append(retrivedRoles, displayName)
+			}
+		}
+
+		if apiErr != nil {
```
Comment (AmitRoushan) [inline cf/commands/user/user-info.go]: apiErr check is at wrong place . I think apiErr check should be after cmd.userRepo call then iterate through user.

File: cf/commands/user/user-info.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+		if cmd.config.IsMinApiVersion("2.21.0") {
+			users, apiErr = cmd.userRepo.ListUsersInOrgForRoleWithNoUAA(cmd.config.OrganizationFields().Guid, role)
+		} else {
+			users, apiErr = cmd.userRepo.ListUsersInOrgForRole(cmd.config.OrganizationFields().Guid, role)
+		}
+
+		for _, user := range users {
+			if cmd.config.Username() == user.Username {
+				retrivedRoles = append(retrivedRoles, displayName)
```
Comment (AmitRoushan) [inline cf/commands/user/user-info.go]: Break for loop once you got user as you are getting all user on role basis. So no need to iterate further.

File: cf/commands/user/user-info.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	var orgRoleToDisplayName = map[string]string{
+		models.ORG_USER:        T("USERS"),
+		models.ORG_MANAGER:     T("ORG MANAGER"),
+		models.BILLING_MANAGER: T("BILLING MANAGER"),
+		models.ORG_AUDITOR:     T("ORG AUDITOR"),
+	}
+	retrivedRoles := []string{} //Variable to store all the roles
+	var users []models.UserFields
+	var apiErr error
+	var roles = []string{models.ORG_MANAGER, models.BILLING_MANAGER, models.ORG_AUDITOR}
```
Comment (AmitRoushan) [inline cf/commands/user/user-info.go]: We are not checking roles based on "USERS" , so either we have to use it for querying or we should remove it from  'orgRoleToDisplayName '

File: cf/commands/user/user-info.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	var usersSpace []models.UserFields
+	for _, role := range spaceRoles {
+		displayName := spaceRoleToDisplayName[role]
+
+		if cmd.config.IsMinApiVersion("2.21.0") {
+			usersSpace, apiErr = cmd.userRepo.ListUsersInSpaceForRoleWithNoUAA(cmd.config.SpaceFields().Guid, role)
+		} else {
+			usersSpace, apiErr = cmd.userRepo.ListUsersInSpaceForRole(cmd.config.SpaceFields().Guid, role)
+		}
+
```
Comment (AmitRoushan) [inline cf/commands/user/user-info.go]: Please consider Organization role related comment of putting break in userSpace iteration and apiErr check for space role also

File: cf/commands/user/user-info.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
+package user
+
+import (
+	//"errors"
```
Comment (AmitRoushan) [inline cf/commands/user/user-info.go]: Please remove commented part of code .

File: cf/app/help.go [hunk 6]
```diff
@@ -105,6 +105,7 @@ func newAppPresenter(app *cli.App) (presenter appPresenter) {
 					presentCommand("logout"),
 					presentCommand("passwd"),
 					presentCommand("target"),
+					presentCommand("user-info"),
```
Comment (AmitRoushan) [inline cf/app/help.go]: I think 'user-info' command can not go with 'GETTING STARTED' group of commands. Please check.

File: cf/command_factory/factory.go [hunk 7]
```diff
@@ -124,6 +124,7 @@ func NewFactory(ui terminal.UI, config core_config.ReadWriter, manifestRepo mani
 	factory.cmdsByName["logs"] = application.NewLogs(ui, config, repoLocator.GetLogsRepository())
 	factory.cmdsByName["oauth-token"] = commands.NewOAuthToken(ui, config, repoLocator.GetAuthenticationRepository())
 	factory.cmdsByName["org"] = organization.NewShowOrg(ui, config)
+	factory.cmdsByName["user-info"] = user.ShowUserInfo(ui, config, repoLocator.GetUserRepository())
```
Comment (AmitRoushan) [inline cf/command_factory/factory.go]: I think better to put "user-info" command registration line after 'update-security-group' command registration as most of commands are in alphabetical order.

**Comment by AmitRoushan** [comment]
@asifhuawei  @SrinivasChilveri 
I agree with SrinivasChilveri  , We can optimize this implementation as CC has given feasibility to get User role based on user ID . With this , We can avoid 'for loop' iteration over user.
Please check -
/v2/users/guid/managed_organizations
/v2/users/guid/billing_managed_organizations
/v2/users/guid/audited_organizations
/v2/users/guid/managed_spaces
/v2/users/guid/audited_spaces api endpoint provided by CC

**Comment by asifhuawei** [comment]
@AmitRoushan @SrinivasChilveri 
If I am going to use 
/v2/users/guid/managed_organizations
/v2/users/guid/billing_managed_organizations
/v2/users/guid/audited_organizations
/v2/users/guid/managed_spaces
/v2/users/guid/audited_spaces  
these endpoints then can you give me a sample output of how to show user info in context for this story  https://www.pivotaltracker.com/n/projects/892938/stories/88825822 ...

These end points are going to return all the organization/spaces the user has access to, and how to consolidate that list in the form of the story context.

**Comment by AmitRoushan** [comment]
@asifhuawei  @SrinivasChilveri 

Api endpoints , i have suggested will give organization/space for particular user based on role.
Here we will get advantage as in real time there will we less organization/space as compared to user.

Additionally you have to check other endpoint api which through we can implement it optimally.
For example -
"/v2/organizations?q=manager_guid:16cf1bcb-31bb-483d-ad30-7b2b0794742f\;q=name:Orgname"
You can list organization based on role of user .

Here is sample for that -
organization based on user role

**Comment by asifhuawei** [comment]
@AmitRoushan @SrinivasChilveri 

so as per your suggestion should we display the info like this

```
$ cf user-info
Getting user information......
User    Org       Space          Role
asif   asifOrg1   ----           ORG MANAGER
asif   asifOrg2   ----           Billing MANAGER
asif    ---       asifSpace1   SPACE MANAGER
asif    ---       asifSpace2   SPACE DEVELOPE
asif    ---       asifSpace3   SPACE DEVELOPE

```

????
or is it a better choice to show the Role of the user according to the organization and space its targeted

**Comment by asifhuawei** [comment]
@SrinivasChilveri  @AmitRoushan  @shashidharatd  @irfanurrehman 

The changes has been done as per review comments, please review, right now there are multiple commits in this PR so once the review is done I will rebase it to single commit.

File: cf/commands/user/user-info.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	var orgName = cmd.config.OrganizationFields().Name
+
+	var orgRoleToDisplayName = map[string]string{
+		models.ORG_MANAGER:     T("ORG MANAGER"),
+		models.BILLING_MANAGER: T("BILLING MANAGER"),
+		models.ORG_AUDITOR:     T("ORG AUDITOR"),
+	}
+
+	for apiName, displayName := range models.ApiToOrgRole {
+		orgs, apiErr := cmd.orgRepo.GetRole(apiName, orgName)
```
Comment (AmitRoushan) [inline cf/commands/user/user-info.go]: Please provide meaningful name for 'GetRole' function.

File: cf/commands/user/user-info.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+			cmd.ui.Failed(T("Failed fetching org-users for role {{.OrgRoleToDisplayName}}.\n{{.Error}}",
+				map[string]interface{}{
+					"Error":                apiErr.Error(),
+					"OrgRoleToDisplayName": orgRoleToDisplayName[displayName],
+				}))
+			return
+		}
+
+		for _, org := range orgs {
+			if org.Name != "" {
```
Comment (AmitRoushan) [inline cf/commands/user/user-info.go]: I think 'orgs' will only at most one element . So no need for using 'for loop'.

File: cf/commands/user/user-info.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+				retrivedRoles = append(retrivedRoles, orgRoleToDisplayName[displayName])
+				break
+			}
+		}
+	}
+
+	// Fetching space roles
+	if cmd.config.SpaceFields().Name != "" {
+		var spaceRoleToDisplayName = map[string]string{
+			models.SPACE_MANAGER:   T("SPACE MANAGER"),
```
Comment (AmitRoushan) [inline cf/commands/user/user-info.go]: I suggest if we are not modifying 'spaceRoleToDisplayName ' and 'orgRoleToDisplayName ' we can declare them as constant.

File: cf/commands/user/user-info_test.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+		return testcmd.RunCommand(ShowUserInfo(ui, configRepo, orgRepo, spaceRepo), args, requirementsFactory)
+	}
+
+	Describe("requirements", func() {
+		It("fails when not logged in", func() {
+			Expect(runCommand()).To(BeFalse())
+		})
+
+		It("fails when no org or space is targeted", func() {
+			requirementsFactory.LoginSuccess = true
```
Comment (AmitRoushan) [inline cf/commands/user/user-info_test.go]: Please rewrite this testcase as now we are only expecting ORG to be targeted .

File: cf/commands/user/user-info_test.go [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+			runCommand()
+
+			Expect(ui.Outputs).NotTo(ContainSubstrings(
+				[]string{"SPACE MANAGER", "SPACE DEVELOPER"},
+			))
+		})
+
+	})
+
```
Comment (AmitRoushan) [inline cf/commands/user/user-info_test.go]: Testcases for only targeted Org are missing.

File: cf/commands/user/user-info_test.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+				[]string{"user", "Org1", "Space1", "SPACE MANAGER", "SPACE DEVELOPER"},
+			))
+		})
+
+		It("showing no roles for  orgs and spaces", func() {
+
+			runCommand()
+
+			Expect(ui.Outputs).NotTo(ContainSubstrings(
+				[]string{"SPACE MANAGER", "SPACE DEVELOPER"},
```
Comment (AmitRoushan) [inline cf/commands/user/user-info_test.go]: Please add roles for ORG and SPACE in expectation.

File: cf/commands/user/user-info.go [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/cloudfoundry/cli/cf/api/spaces"
+	"github.com/cloudfoundry/cli/cf/command_metadata"
+	"github.com/cloudfoundry/cli/cf/configuration/core_config"
+	. "github.com/cloudfoundry/cli/cf/i18n"
+	"github.com/cloudfoundry/cli/cf/models"
+	"github.com/cloudfoundry/cli/cf/requirements"
+	"github.com/cloudfoundry/cli/cf/terminal"
+	"github.com/codegangsta/cli"
+	"strings"
+)
```
Comment (AmitRoushan) [inline cf/commands/user/user-info.go]: 'strings' package should be at top of package import as per convention .

File: cf/models/user_roles.go [hunk 15]
```diff
@@ -27,3 +27,15 @@ var SpaceRoleToUserInput = map[string]string{
...
+	"manager_guid":         ORG_MANAGER,
+	"billing_manager_guid": BILLING_MANAGER,
+	"auditor_guid":         ORG_AUDITOR,
+}
+
+var ApiToSpaceRole = map[string]string{
+	"managed_spaces": SPACE_MANAGER,
+	"developer_guid": SPACE_DEVELOPER,
+	"audited_spaces": SPACE_AUDITOR,
+}
```
Comment (AmitRoushan) [inline cf/models/user_roles.go]: Suggestion -
We can rename ApiToOrgRole  as QueryParmToOrgRole and ApiToSpaceRole  with QueryParmToSpaceRole

**Comment by asifhuawei** [comment]
@SrinivasChilveri @AmitRoushan @shashidharatd @irfanurrehman

The changes has been done as per review comments, please review......

**Pull request closed without merging**